msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java

```diff
@@ -554,51 +557,127 @@ public void run() {
+        if (MSALControllerFactory.brokerEligible(
+                mPublicClientConfiguration.getAppContext(),
+                mPublicClientConfiguration.getDefaultAuthority(),
+                mPublicClientConfiguration)) {
+            com.microsoft.identity.common.internal.logging.Logger.info(
+                    TAG + methodName,
+                    "Eligible to use broker. Try to get the account from local and broker cache."
+            );
+            getBrokerAndLocalAccount(oid, callback);
         } else {
-            com.microsoft.identity.common.internal.logging.Logger.warn(
+            // load local account and find the matching account
+            com.microsoft.identity.common.internal.logging.Logger.info(
                     TAG + methodName,
-                    "Provided authority was not AAD - defaulting to parsed home_account_id"
+                    "Not eligible to use broker. Try to get the account from local cache."
             );
+            getLocalAccount(oid, callback);
         }
+    }
 
-        AccountRecord accountToReturn = null;
+    private void getBrokerAndLocalAccount(@NonNull final String oid,
+                                          @NonNull final AccountsLoadedCallback<IAccount> callback) {
+        final String methodName = ":getBrokerAndLocalAccount";
 
-        if (null != realm) {
-            accountToReturn = AccountAdapter.getAccountInternal(
-                    mPublicClientConfiguration.getClientId(),
-                    mPublicClientConfiguration.getOAuth2TokenCache(),
-                    homeAccountIdentifier,
-                    realm
-            );
-        } else {
-            com.microsoft.identity.common.internal.logging.Logger.warn(
+        final Handler handler = getPreferredLooper();
+
+        //Make a background call to get account from broker
+        new BrokerMsalController().getBrokerAccounts(
+                mPublicClientConfiguration,
+                new BrokerAccountsLoadedCallback() {
+                    @Override
+                    public void onAccountsLoaded(final List<AccountRecord> accountRecords) {
+                        final List<AccountRecord> cachedAccounts = new ArrayList<>();
+                        cachedAccounts.addAll(getLocalAccounts());
+                        cachedAccounts.addAll(accountRecords);
+
+                        if (cachedAccounts.isEmpty()) {
+                            com.microsoft.identity.common.internal.logging.Logger.info(
+                                    TAG + methodName,
+                                    "No account stored in the broker cache."
+                            );
+                            handler.post(new Runnable() {
+                                @Override
+                                public void run() {
+                                    //Return null if no account found.
+                                    callback.onAccountsLoaded(null);
```

Comment:
Let's return an empty `List` instead - otherwise we force callers to check for both `null` and `isEmpty()`

Reply:
This function is trying to find the target account, the one with matching oid, from local msal and broker's cache. Not the entire account list.